File: Data Structures & Algorithms/daily-temperatures/submission-0.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        ArrayDeque<int[]> stack = new ArrayDeque<>(); //[idx, temp]
        for (int i = 0; i < temperatures.length; i++) {
            int temp = temperatures[i];
            while (!stack.isEmpty() && temp > stack.peekLast()[1]) {
                int tailIdx = stack.removeLast()[0];
                res[tailIdx] = i - tailIdx;
            }
            stack.add(new int[]{i, temp});
        }
        return res;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1]*len(nums)
        p = 1
        for i in range(len(nums)):
            if i > 0:
                res[i] = p
            p = p * nums[i]
        
        p = 1
        for i in range(len(nums) - 1, -1, -1):
            if (i < len(nums) - 1):
                res[i] = res[i] * p
            p = p * nums[i]
        
        return res